dashboard/ltiv1p3.py

```diff
@@ -0,0 +1,144 @@
+    return tool_config
+
+
+def get_launch_url(request):
+    target_link_uri = request.POST.get('target_link_uri', request.GET.get('target_link_uri'))
+    if not target_link_uri:
+        raise Exception('Missing "target_link_uri" param')
+    return target_link_uri
+
+
+def get_jwks(request):
+    logger.info("************** get_jwks() ")
+    tool_conf = get_tool_conf()
+    return JsonResponse(tool_conf.get_jwks(), safe=False)
+
+
+def check_if_instructor(roles, username, course_id):
+    teaching_roles = [role for role in roles if "membership#Instructor" in role or
+                      "membership/Instructor#TeachingAssistant" in role]
+    if len(teaching_roles) == 0:
+        logger.info(f"user {username} is not instructor in course {course_id}")
+        return False
+
+    ta = [teach_role for teach_role in teaching_roles if "membership/Instructor#TeachingAssistant" in teach_role]
+    if len(ta) != 0:
+        logger.info(f"user {username} is not a instructor of course {course_id}")
+        return False
+    else:
+        logger.info(f"user {username} is instructor in course {course_id}")
+        return True
+
+
+def validate_lti_1_3_request(request, message_launch):
+    launch_id = message_launch.get_launch_id()
+    launch_data = message_launch.get_launch_data()
+    custom_params = launch_data['https://purl.imsglobal.org/spec/lti/claim/custom']
+    course_name = launch_data['https://purl.imsglobal.org/spec/lti/claim/context']['title']
+    roles = launch_data['https://purl.imsglobal.org/spec/lti/claim/roles']
+    username = custom_params['user_username']
+    course_id = custom_params['canvas_course_id']
+    email = launch_data['email']
+    first_name = launch_data['given_name']
+    last_name = launch_data['family_name']
+    user_img = launch_data['picture']
+    RANDOM_PASSWORD_DEFAULT_LENGTH = 32
+
+    # Logging the user regardless course exits or not otherwise Django will show default login page
+    try:
+        user_obj = User.objects.get(username=username)
+    except User.DoesNotExist:
+        password = ''.join(random.sample(string.ascii_letters, RANDOM_PASSWORD_DEFAULT_LENGTH))
+        user_obj = User.objects.create_user(username=username, email=email, password=password, first_name=first_name, last_name=last_name)
+    user_obj.backend = 'django.contrib.auth.backends.ModelBackend'
+    django.contrib.auth.login(request, user_obj)
+
+    is_instructor = check_if_instructor(roles, username, course_id)
+
+    try:
+        course_details = Course.objects.get(canvas_id=course_id)
+    except ObjectDoesNotExist:
+        if is_instructor:
+            canvas_long_id = canvas_id_to_incremented_id(course_id)
+            course_details = Course.objects.create(id=canvas_long_id, canvas_id=course_id, name=course_name)
+            CourseViewOption.objects.create(course_id=canvas_long_id)
+
+    if course_details.term_id is None:
+        logger.info(f"Course {course_id} don't have a cron run yet")
```

Comment:
I'd be a little less specific.  That is, don't mention `cron` specifically.  If the job is run by Jenkins, that doesn't use `cron` at all.  Maybe something like…

```suggestion
        logger.info(f'Course ({course_id}) data loading job has not run yet.')
```

Yeah, I know it's just a log statement, but it could be used as a user-facing message, too.

I see a lot of mixing of single and double quoting in the code.  Best practices say to be consistent.  My advice is to use single quotes in most places and double quotes only when they are necessary.